frontend/src/service-worker.js

```diff
@@ -1,15 +1,72 @@
-import {serviceWorkerFetchListener} from "sync-message";
+/* eslint-disable */
+// Otherwise webpack can fail silently
+// https://github.com/facebook/create-react-app/issues/8014
 
-console.log(self.__WB_MANIFEST);
+// This service worker can be customized!
+// See https://developers.google.com/web/tools/workbox/modules
+// for the list of available Workbox modules, or add any other
+// code you'd like.
+// You can also remove this file if you'd prefer not to use a
+// service worker, and the Workbox build step will be skipped.
 
-const fetchListener = serviceWorkerFetchListener();
+import { clientsClaim } from 'workbox-core';
+import { ExpirationPlugin } from 'workbox-expiration';
+import { precacheAndRoute, createHandlerBoundToURL } from 'workbox-precaching';
+import { registerRoute } from 'workbox-routing';
+import { StaleWhileRevalidate } from 'workbox-strategies';
 
-addEventListener('fetch', fetchListener);
+clientsClaim();
 
-addEventListener('install', function (e) {
-  e.waitUntil(self.skipWaiting());
-});
+// Precache all of the assets generated by your build process.
+// Their URLs are injected into the manifest variable below.
+// This variable must be present somewhere in your service worker file,
+// even if you decide not to use precaching. See https://cra.link/PWA
+precacheAndRoute(self.__WB_MANIFEST);
+
+// Set up App Shell-style routing, so that all navigation requests
+// are fulfilled with your index.html shell. Learn more at
+// https://developers.google.com/web/fundamentals/architecture/app-shell
+const fileExtensionRegexp = new RegExp('/[^/?]+\\.[^/]+$');
+registerRoute(
+  // Return false to exempt requests from being fulfilled by index.html.
+  ({ request, url }) => {
+    // If this isn't a navigation, skip.
+    if (request.mode !== 'navigate') {
+      return false;
+    } // If this is a URL that starts with /_, skip.
+
+    if (url.pathname.startsWith('/_')) {
+      return false;
+    } // If this looks like a URL for a resource, because it contains // a file extension, skip.
+
+    if (url.pathname.match(fileExtensionRegexp)) {
+      return false;
+    } // Return true to signal that we want to use the handler.
+
+    return true;
+  },
+  createHandlerBoundToURL(process.env.PUBLIC_URL + '/index.html')
+);
+
+// An example runtime caching route for requests that aren't handled by the
+// precache, in this case same-origin .png requests like those from in public/
+registerRoute(
+  // Add in any other file extensions or routing criteria as needed.
+  ({ url }) => true, // XXX Caching everything for now. It breaks firebase, but it makes most of the PWA work offline. We can narrow it down later.
```

Comment:
Needs to be narrowed down now. Bootstrap should probably be moved into futurecoder rather than served from a CDN. Then that just leaves pyodide.

I'd still like to understand:

1. Why caching from other origins works at all, and if it can be relied upon.
2. Whether it would be good to have precaching for the pyodide stuff anyway

Reply:
OK, now there's more specific rules in `service-worker.js`, and they're only applied when it's a build (i.e. `NODE_ENV=production`).

However, when testing offline, it errors out on the auth/firebase stuff, as we were seeing before. So, offline-readiness might be blocked until changes there.